### ISSUE TITLE
Inconsistent behavior with circular import on MacOS

### ISSUE THREAD

**Issue by guyarad** [pylint-dev/pylint#9168]
### Bug description

```python
The issue appeared when the code allegedly had a circular dependency. `pylint` did not complained on a MacOS machine, but did complain on Linux. However, I was able to narrow it down to a file-system (???) issue by demonstrating an inconsistent behavior when running inside a container on a local folder versus a mounted folder (when the host is a MacOS).

I added details and how-to reproduce [here](https://github.com/guyarad/pylint-macos-issue)
```

### Command used

```shell
pylint --recursive=y --disable=W,C0114,C0115,C0116,R0903 -s=n .
```

### Pylint output

```python
Inconsistent - on Linux:

************* Module module1.derived
module1/derived.py:1:0: R0401: Cyclic import (module1 -> module1.derived) (cyclic-import)

But on a MacOS it was successful.
```

### Expected behavior

Consistent behavior when running on a mounted folder and local folder (when all files are identical).

### Pylint version

```shell
pylint 3.0.1
astroid 3.0.1
Python 3.11.6 (main, Oct  3 2023, 03:39:03) [GCC 10.2.1 20210110]
```

### OS / Environment

  Model Name:	MacBook Pro
  Model Identifier:	Mac14,10
  Chip:	Apple M2 Pro

OS Version: Ventura 13.5

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue. I created a reproducer at #11048, and the macOS 3.10 job emits nothing at all, so on macOS pylint is silently skipping the entire module1 package's analysis, not just the cyclic-import check. Troubleshooting on MAC is impossible for me, so contribution welcome.
